contracts/upgradeable_contracts/ChaiConnector.sol

```diff
@@ -0,0 +1,191 @@
+    * @param recipient Account address to receive available interest
+    */
+    function _payInterest(address recipient) internal {
+        // since investedAmountInChai() returns a ceiled value,
+        // the value of chaiBalance() - investedAmountInChai() will be floored,
+        // leading to excess remaining chai balance
+        uint256 balanceBefore = erc20token().balanceOf(address(this));
+        chaiToken().exit(address(this), chaiBalance().sub(investedAmountInChai()));
+        uint256 interestInDai = erc20token().balanceOf(address(this)) - balanceBefore;
+
+        erc20token().transfer(recipient, interestInDai);
+
+        require(dsrBalance() >= investedAmountInDai());
+    }
+
+    /**
+    * @dev Evaluates bridge balance for tokens, holded in DSR
+    * @return Balance in dai, truncated
+    */
+    function dsrBalance() public view returns (uint256) {
+        uint256 chi = chaiToken().pot().chi();
+        return rmul(chi, chaiBalance());
+    }
+
+    /**
+    * @dev Evaluates bridge balance in Chai tokens
+    * @return Balance in chai, exact
+    */
+    function chaiBalance() public view returns (uint256) {
+        return chaiToken().balanceOf(address(this));
+    }
+
+    /**
+    * @dev Evaluates exact current invested amount, id DAI
+    * @return Value in DAI
+    */
+    function investedAmountInDai() public view returns (uint256) {
+        return uintStorage[INVESTED_AMOUNT];
+    }
+
+    /**
+    * @dev Updates current invested amount, id DAI
+    * @return Value in DAI
+    */
+    function setInvestedAmointInDai(uint256 amount) internal {
+        uintStorage[INVESTED_AMOUNT] = amount;
+    }
+
+    /**
+    * @dev Evaluates amount of chai tokens that is sufficent to cover 100% of the invested DAI
+    * @return Amount in chai, ceiled
+    */
+    function investedAmountInChai() internal returns (uint256) {
+        IPot pot = chaiToken().pot();
+        // solhint-disable-next-line not-rely-on-time
+        uint256 chi = (now > pot.rho()) ? pot.drip() : pot.chi();
+        return rdivup(investedAmountInDai(), chi);
+    }
+
+    /**
+    * @dev Invests DAI into Chai
+    * @param amount Amount of DAI to invest
+    */
+    function _convertDaiToChai(uint256 amount) internal {
+        setInvestedAmointInDai(investedAmountInDai() + amount);
+        erc20token().approve(chaiToken(), amount);
+        chaiToken().join(address(this), amount);
+    }
+
+    /**
+    * @dev Redeems DAI from Chai, the total redeemed amount will be at least equal to specified amount
+    * @param amount Amount of DAI to redeem
+    */
+    function _convertChaiToDai(uint256 amount) internal {
+        uint256 invested = investedAmountInDai();
+        if (amount >= invested) {
+            // onExecuteMessage can call a convert operation with argument greater than the current invested amount,
+            // in this case bridge should withdraw all invested funds
+            chaiToken().draw(address(this), invested);
```

Comment:
The question about Chai/DSR: what is more profitable: pay the interest before Dai withdrawal or continue to keep the interest until the method `payInterest` is called?

Reply:
There is no any profit in paying interest before Dai withdrawal. 
In DSR, after each second the interest is reevaluated for your account. The formula depends only on the current amount of DAI saved in DSR. So there is no need to explicitly
call `payInterest` before withdrawal.
The optimal way is to keep DAI in DSR as long as you can, and then pay the interest only when it is needed.

Reply:
Ok. Thanks for clarification.